### ISSUE TITLE
Sometimes using inplace=True with Relu causes an error

### ISSUE THREAD

**Issue by rhewett** [distdl/distdl#199]
For example, in U-nets, we are getting an error:

???

This error is fixed by setting `inplace=False` for the relu applications.  This does not happen on 0.3.1 but does for 0.4-dev.

**Comment by rhewett** [comment]
After some digging:
* Removing `ctx.mark_dirty(input)` in the forward halo exchange function causes the error to go away.  But it does not make sense to remove this as the input *is* modified in place and thus becomes dirty.
* Halo exchange tests still pass after removing `ctx.mark_dirty(input)`
* With `ctx.mark_dirty(input)`, checking the `id()` verifies that the data is modified in place.  Without it, this happens sometimes.  In this test, the first call is performed in place but later calls are not.  This does not make sense.
* PyTorch documentation seems to indicate that `inplace=True`  only results in minimal memory savings.
* It is completely opaque to me as to why the changes to the padding structure made between v0.3 and v0.4 has caused this issue to materialize.

Until this is resolved, given that repeated halo calls can be very expensive from a memory perspective, my suggestion is to use `inplace=False` on the ReLUs.  This certainly demands further investigation.

**Comment by rhewett** [comment]
FWIW, this code should allow the issue to be replicated but it actually works correctly afaik:

`
import torch

class CustomFunc(torch.autograd.Function):

    @staticmethod
    def forward(ctx, input):

        # ctx.mark_dirty(input)

        # v = input.detach().numpy()
        # v[2] = 4

        return input

    @staticmethod
    def backward(ctx, grad_output):

        # grad_input = 0*grad_output

        return torch.ones(5)

class CustomLayer(torch.nn.Module):

    def forward(self, input):

        return CustomFunc.apply(input)

class NotIdentity(torch.nn.Module):

    def forward(self, input):
        return 2*input

net2 = torch.nn.Sequential(NotIdentity(),
                           CustomLayer(),
                           torch.nn.ReLU(inplace=True))

x2 = torch.ones(5)
x2.requires_grad = True
dy2 = torch.ones(5)

y2 = net2(x2)
y2.backward(dy2)
dx2 = x2.grad
`
